### PR TITLE
IGNITE-20452 Integrate failure handler processor into Ignite components

### PR DESCRIPTION
**Pull request by sk0x50**
https://issues.apache.org/jira/browse/IGNITE-20452

The failure processor was propagated into the following Ignite components:
 - network
 - page-memory
 - watchdog
 - tables

Also , this pull request includes the failure handler feature introduced by https://issues.apache.org/jira/browse/IGNITE-20447 and https://issues.apache.org/jira/browse/IGNITE-20448 from the feature branch.

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/failure-handler/src/main/java/org/apache/ignite/internal/failure/FailureContext.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+     * @param type Failure type.
+     * @param err Exception.
+     */
+    public FailureContext(FailureType type, Throwable err) {
+        this.type = type;
+        this.err = err;
+    }
+
+    /**
+     * Gets the failure type.
```
Comment (ibessonov) [inline modules/failure-handler/src/main/java/org/apache/ignite/internal/failure/FailureContext.java:45]: Can I ask you to replace all these ridiculous "gets" with "returns"?
I understand that it's a copy of a pre-existing class, but it looks weird nonetheless.

File: modules/failure-handler/src/main/java/org/apache/ignite/internal/failure/FailureContext.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    /**
+     * Gets the exception.
+     *
+     * @return Exception or {@code null}.
+     */
+    public Throwable error() {
+        return err;
+    }
+
+    /** {@inheritDoc} */
```
Comment (ibessonov) [inline modules/failure-handler/src/main/java/org/apache/ignite/internal/failure/FailureContext.java:62]: This is also not necessary, but I don't mind it

File: modules/failure-handler/src/main/java/org/apache/ignite/internal/failure/handlers/StopNodeOrHaltFailureHandler.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+            ).start();
+        } else {
+            Runtime.getRuntime().halt(KILL_EXIT_CODE);
+        }
+
+        return true;
+    }
+
+    /**
+     * Get stop node timeout in milliseconds.
```
Comment (ibessonov) [inline modules/failure-handler/src/main/java/org/apache/ignite/internal/failure/handlers/StopNodeOrHaltFailureHandler.java:97]: Please correct these javadocs as well

File: modules/failure-handler/src/test/java/org/apache/ignite/internal/failure/FailureProcessorTest.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    @Test
+    void testFailureProcessing() {
+        FailureHandler handler = mock(FailureHandler.class);
+
+        FailureProcessor failureProcessor = new FailureProcessor("node_name", handler);
+
+        assertThat(failureProcessor.start(), willSucceedFast());
+
+        try {
+
```
Comment (ibessonov) [inline modules/failure-handler/src/test/java/org/apache/ignite/internal/failure/FailureProcessorTest.java:45]:
```suggestion
```

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/checkpoint/CheckpointReadLockTimeoutException.java [hunk 5]
```diff
@@ -32,6 +33,6 @@ public class CheckpointReadLockTimeoutException extends IgniteInternalException
      * @param msg Error message.
      */
     CheckpointReadLockTimeoutException(String msg) {
-        super(msg);
+        super(SYSTEM_CRITICAL_OPERATION_TIMEOUT_ERR, msg);
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/checkpoint/CheckpointReadLockTimeoutException.java:36]: Not sure about that, to be honest, but better safe than sorry, right?

File: modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java [hunk 6]
```diff
@@ -390,11 +394,15 @@ public class IgniteImpl implements Ignite {
 
         MessageSerializationRegistry serializationRegistry = createSerializationRegistry(serviceProviderClassLoader);
 
+        // TODO https://issues.apache.org/jira/browse/IGNITE-20450
+        failureProcessor = new FailureProcessor(name);
```
Comment (ibessonov) [inline modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java:398]: Who's calling `start` and `stop` for this component? Is this done automatically?

File: modules/api/src/main/java/org/apache/ignite/lang/ErrorGroups.java [hunk 7]
```diff
@@ -580,4 +580,17 @@ public static class PlacementDriver {
         /** Primary replica await error. */
         public static final int PRIMARY_REPLICA_AWAIT_ERR = PLACEMENT_DRIVER_ERR_GROUP.registerErrorCode((short) 2);
     }
+
+    /** Critical workers error group. */
+    @ErrorCodeGroup
+    public static class CriticalWorkers {
```
[the remark this answers is not in the document]
Comment (sk0x50) [reply modules/api/src/main/java/org/apache/ignite/lang/ErrorGroups.java:586]: It seems to me that using a specific error code is better than using a general internal error even though a node failed due to a blocked critical worker.

File: modules/failure-handler/src/main/java/org/apache/ignite/internal/failure/FailureContext.java (hunk 1 above)
Comment (sk0x50) [reply modules/failure-handler/src/main/java/org/apache/ignite/internal/failure/FailureContext.java:45]: makes sense to me. will do.

File: modules/failure-handler/src/main/java/org/apache/ignite/internal/failure/handlers/StopNodeOrHaltFailureHandler.java (hunk 3 above)
Comment (sk0x50) [reply modules/failure-handler/src/main/java/org/apache/ignite/internal/failure/handlers/StopNodeOrHaltFailureHandler.java:97]: Fixed

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/checkpoint/CheckpointReadLockTimeoutException.java (hunk 5 above)
Comment (sk0x50) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/checkpoint/CheckpointReadLockTimeoutException.java:36]: Yep, I think that the more concrete the error code the better.

File: modules/failure-handler/src/main/java/org/apache/ignite/internal/failure/FailureContext.java (hunk 2 above)
Comment (sk0x50) [reply modules/failure-handler/src/main/java/org/apache/ignite/internal/failure/FailureContext.java:62]: Removed.

File: modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java (hunk 6 above)
Comment (sk0x50) [reply modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java:398]: Nice catch! Fixed.

**Review by ibessonov — APPROVED** [review]
